lib/OpenQA/CacheService/Controller/API.pm

```diff
@@ -32,62 +24,36 @@ sub info {
 sub status {
     my $self = shift;
 
-    my $data      = $self->req->json;
-    my $lock_name = $data->{lock};
-    return $self->render(json => {error => 'No lock specified.'}, status => 400) unless $lock_name;
-
-    my $lock = $self->gen_guard_name($lock_name);
-    my %res  = (
-        status => (
-              $self->_active($lock)         ? STATUS_DOWNLOADING
-            : $self->locks->enqueued($lock) ? STATUS_IGNORE
-            :                                 STATUS_PROCESSED
-        ));
+    my $data = $self->req->json;
+    return $self->render(json => {error => 'No job specified'}, status => 400) unless my $id = $data->{id};
 
-    if ($data->{id}) {
-        my $job = $self->minion->job($data->{id});
-        return $self->render(json => {error => 'Specified job ID is invalid.'}, status => 404) unless $job;
+    return $self->render(json => {error => 'Specified job ID is invalid'}, status => 404)
+      unless my $job = $self->minion->job($id);
+    return $self->render(json => {error => 'Job info not available'}, status => 404)
+      unless my $job_info = $job->info;
 
-        my $job_info = $job->info;
-        return $self->render(json => {error => 'Job info not available.'}, status => 404) unless $job_info;
-        $res{result} = $job_info->{result};
-        $res{output} = $job_info->{notes}->{output};
-    }
+    # The job will finish early if another job is already downloading, so we have to wait for the lock too
+    $self->render(json => {status => 'processed', result => $job_info->{result}, output => $job_info->{notes}{output}})
+      if $job_info->{state} eq 'finished' && !$self->progress->is_downloading($job_info->{notes}{lock});
```

Comment:
I'm unclear... how is this `render` call waiting for the lock? I might be misunderstanding the comment, or maybe I'm not clear what the code is meant to do.

Reply:
The comment is about the `if $job_info->{state} eq 'finished' && !$self->progress->is_downloading($job_info->{notes}{lock})` check. If there is already another background job downloading the asset, our job will finish immediately. And then we have to wait for the lock to be freed by the other background job (by calling `->progress->is_downloading`). This action (`/status`) gets called over and over until the status in the JSON response is `processed`.

Reply:
Feel free to suggest a better explanation, i'd like for this to be as simple to understand as possible. That's why i'm making all these changes in the first place. 😉

Reply:
Right. Thanks for clarifying.

I think it's the use of "we" and "wait" here that makes me think of blocking on the result which isn't really what's happening and the status call isn't doing anything in the first place.

The job will wait on the lock. And then the real consumer will take it from there.

Reply:
It's very possible that I'm getting lost in semantics. But from what I see in the code, the status call is not much more than an API doing no actual work.

Reply:
I'll turn this route into a `GET /status/<id>`, that should make its nature more obvious.

Reply:
Even better :+1: